tango/step_caches/remote_step_cache.py

```diff
@@ -0,0 +1,199 @@
+    RemoteDataset,
+    RemoteDatasetConflict,
+    RemoteDatasetNotFound,
+    RemoteDatasetWriteError,
+)
+from tango.step import Step
+from tango.step_cache import CacheMetadata
+from tango.step_caches.local_step_cache import LocalStepCache
+from tango.step_info import StepInfo
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteNotFoundError(TangoError):
+    """
+    Classes inheriting from the RemoteStepCache should raise this if a step result object is not found.
+    """
+
+
+# This class inherits from `LocalStepCache` to benefit from its in-memory "weak cache" and "strong cache",
+# but it handles saving artifacts to disk a little differently.
+class RemoteStepCache(LocalStepCache):
+    """
+    This is a :class:`~tango.step_cache.StepCache` that's used by :class:`RemoteWorkspace`.
+    It stores the results of steps on some RemoteWorkspace.
+
+    It also keeps a limited in-memory cache as well as a local backup on disk, so fetching a
+    step's resulting subsequent times should be fast.
+
+    .. tip::
+        All remote step caches inherit from this.
+    """
+
+    Constants = RemoteConstants
+
+    def __init__(self, local_dir: Path):
+        super().__init__(local_dir)
+
+    @property
+    @abstractmethod
+    def client(self) -> RemoteClient:
+        raise NotImplementedError()
+
+    def _step_result_remote(self, step: Union[Step, StepInfo]) -> Optional[RemoteDataset]:
+        """ """
+        try:
+            dataset = self.client.get(self.Constants.step_dataset_name(step))
+            return dataset if dataset.committed else None
+        except RemoteDatasetNotFound:
+            return None
+
+    def _sync_step_remote(self, step: Step, objects_dir: Path) -> RemoteDataset:
+        dataset_name = self.Constants.step_dataset_name(step)
+        try:
+            self.client.create(dataset_name, commit=False)
+        except RemoteDatasetConflict:
+            pass
+        try:
+            self.client.sync(dataset_name, objects_dir)
+            self.client.commit(dataset_name)
+        except RemoteDatasetWriteError:
+            pass
+
+        return self.client.get(dataset_name)
+
+    def _fetch_step_remote(self, step_result, target_dir: PathOrStr) -> None:
+        try:
+            self.client.fetch(step_result, target_dir)
+        except RemoteDatasetNotFound:
+            self._raise_remote_not_found()
+
+    def __len__(self):
+        # NOTE: lock datasets should not count here. They start with the same prefix,
+        # but they never get committed.
+        return sum(
+            1
+            for ds in self.client.datasets(
+                uncommitted=False, match=self.Constants.STEP_DATASET_PREFIX
+            )
+            if ds.name is not None and ds.name.startswith(self.Constants.STEP_DATASET_PREFIX)
+        )
+
+    def _step_results_dir(self) -> str:
+        return self.Constants.STEP_RESULT_DIR
+
+    def _raise_remote_not_found(self):
+        raise RemoteNotFoundError()
```

Comment:
What are these for? The functions do almost nothing, and don't add any clarity to the code either.

Reply:
Yeah, leftover from when I was trying to build the abstractions.